src/libtls/tls_hkdf.c

```diff
@@ -0,0 +1,520 @@
+	{
+		DBG1(DBG_TLS, "%N not supported", hash_algorithm_names, this->hash_algorithm);
+		return FALSE;
+	}
+
+	DBG3(DBG_TLS, "HASH: %B", &context);
+
+	expand_label(this, &this->prk, label, &context, this->L, &this->okm);
+
+	chunk_free(&context);
+
+	return TRUE;
+}
+
+static bool move_to_phase_1(private_tls_hkdf_t *this)
+{
+	switch (this->phase)
+	{
+		case HKDF_PHASE_0:
+			if (!extract(this, &this->salt, &this->ikm, &this->prk))
+			{
+				DBG1(DBG_TLS, "unable extract PRK");
+				return FALSE;
+			}
+			this->phase = HKDF_PHASE_1;
+			return TRUE;
+		case HKDF_PHASE_1:
+			return TRUE;
+		default:
+			DBG1(DBG_TLS, "invalid HKDF phase");
+			return FALSE;
+	}
+}
+
+static bool move_to_phase_2(private_tls_hkdf_t *this)
+{
+	chunk_t empty, derived;
+
+	switch (this->phase)
+	{
+		case HKDF_PHASE_0:
+			move_to_phase_1(this);
+		case HKDF_PHASE_1:
+			empty = chunk_empty;
+			derived = chunk_from_str("tls13 derived");
+			if (!derive_secret(this, &derived, &empty))
+			{
+				DBG1(DBG_TLS, "unable to derive secret");
+				return FALSE;
+			}
+
+			if (!this->shared_secret)
+			{
+				DBG1(DBG_TLS, "no shared secret set");
+				return FALSE;
+			}
+			else
+			{
+				chunk_free(&this->ikm);
+				this->ikm = chunk_alloc(this->shared_secret->len);
+				memcpy(this->ikm.ptr, this->shared_secret->ptr, this->shared_secret->len);
+				this->ikm.len = this->shared_secret->len;
+			}
+
+			if (!extract(this, &this->okm, &this->ikm, &this->prk))
+			{
+				DBG1(DBG_TLS, "unable extract PRK");
+				return FALSE;
+			}
+			this->phase = HKDF_PHASE_2;
+			return TRUE;
+		case HKDF_PHASE_2:
+			return TRUE;
+		default:
+			DBG1(DBG_TLS, "invalid HKDF phase");
+			return FALSE;
+	}
+}
+
+static bool move_to_phase_3(private_tls_hkdf_t *this)
+{
+	chunk_t empty, derived;
+
+	switch (this->phase)
+	{
+		case HKDF_PHASE_0:
+		case HKDF_PHASE_1:
+			move_to_phase_2(this);
+		case HKDF_PHASE_2:
+			/* prepare okm for next extract */
+			empty = chunk_empty;
+			derived = chunk_from_str("tls13 derived");
+			if (!derive_secret(this, &derived, &empty))
+			{
+				DBG1(DBG_TLS, "unable to derive secret");
+				return FALSE;
+			}
+
+			chunk_t ikm = chunk_from_chars(
+				0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
+				0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
+				0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
+				0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
+			);
+
+			if (!extract(this, &this->okm, &ikm, &this->prk))
+			{
+				DBG1(DBG_TLS, "unable extract PRK");
+				return FALSE;
+			}
+			this->phase = HKDF_PHASE_3;
+			return TRUE;
+		case HKDF_PHASE_3:
+			return TRUE;
+		default:
+			DBG1(DBG_TLS, "invalid HKDF phase");
+			return FALSE;
+	}
+}
+
+static bool write_key_to_caller_secret(private_tls_hkdf_t *this, chunk_t *key, chunk_t *secret)
+{
+	*secret = chunk_alloc(this->L);
+	memset(secret->ptr, 0, this->L);
+	memcpy(secret->ptr + secret->len - key->len,
+		   key->ptr, key->len);
+
```

Comment:
```suggestion
	chunk_copy_pad(*secret, *key, 0);
```
Key should not be a pointer, though.